### ISSUE TITLE
Track opponent mon ability

### ISSUE THREAD

**Issue by besteon** [besteon/Ironmon-Tracker#3]
When an opponent's ability is activated, reflect that information in the tracker. A mon can potentially have 2 ability slots.

**Comment by besteon** [comment]
Due to Bizhawk only being able to handle a single `event.onmemoryread` event at a time, this is significantly more challenging. Unfortunately, the `AbilityBattleEffects` function is one of the largest in the game and is called often, so `event.onmemoryexecute` with this function fires all the time.

**Comment by UTDZac** [comment]
With the code optimizations coming in, it's possible this issue can be solved partially my manually checking for certain conditions. 

For example, if the opposing Pokemon has any ability that activates when it appearances, the tracker knows the Pokemon has that ability from a game data read and can reveal that information to the player. Similarly, it could get more complex by checking if the player uses a move like earthquake that should have hit the opponent, but didn't, hence Levitate ability.

Thought ultimately it may be possible to gain a ton of this information and more through ``gBattleCommunication`` data in memory, just have to figure out how to parse it.

**Comment by UTDZac** [comment]
I've managed to find a workaround for showing many of the important abilities that activate in the middle of combat without using on-events. This is done by manually checking the battlescript instruction values intermittently and watching for memory address values that match one of the abilities.

For example, when Drizzle activates, the `gBattlescriptCurrInstr` memory value is set to `0x081d927f`, which is the address `BattleScript_DrizzleActivates`. Thus we can track this ability.

A few issues so far:
1. Not all ability triggers are easy to match. Ones like Poison Point/Liquid Ooze don't have a clear BattleScript address that is unique to the ability activating, and unfortunately overlaps with other things that can trigger on the same address.
2. The `gBattlescriptCurrInstr` tells us that an ability triggered, but it doesn't tell as which Pokemon's ability triggered it. This is not an issue unless both active Pokemon have the same ability. We don't want to track and reveal information on the opposing Pokemon in this scenario, as it could spoil/reveal hidden information when the player's Pokemon's ability triggers first.
3. Each of the address values returned by `gBattlescriptCurrInstr` are not a clean one-to-one map to the BattleScript addresses defined in the rom dump information. Some are offset by +0x1, +0x3, +0x8, and so on. For example, `BattleScript_FlashFireBoost` doesn't match on it's defined value, but does match on +0x1 for FireRed 1.0 and +0x3 for FireRed 1.1
4. Each version of FR and LG appear to have differences in the offsets used for these values. In the example above, the Flash Fire address is different between FR rev 1.0 and rev 1.1. Therefore, this requires manually checking the values in game for each of the four versions of FRLG, which is a good deal of work. I imagine Emerald addresses are also different.
   - For now, most values I found for FR 1.1 seem to work just fine for other games, but it's worth noting some do not.

All of this auto-ability tracking will be included in the upcoming v0.5.0 release
